kernel/src/fs/overlayfs/fs.rs

```diff
@@ -0,0 +1,1158 @@
+    fn link(&self, old: &Arc<dyn Inode>, name: &str) -> Result<()>;
+    fn unlink(&self, name: &str) -> Result<()>;
+    fn rmdir(&self, name: &str) -> Result<()>;
+    fn lookup(&self, name: &str) -> Result<Arc<dyn Inode>>;
+    fn rename(&self, old_name: &str, target: &Arc<dyn Inode>, new_name: &str) -> Result<()>;
+    fn read_link(&self) -> Result<String>;
+    fn write_link(&self, target: &str) -> Result<()>;
+    fn ioctl(&self, cmd: IoctlCmd, arg: usize) -> Result<i32>;
+    fn sync_all(&self) -> Result<()>;
+    fn sync_data(&self) -> Result<()>;
+    fn fallocate(&self, mode: FallocMode, offset: usize, len: usize) -> Result<()>;
+    fn fs(&self) -> Arc<dyn FileSystem>;
+    fn set_xattr(
+        &self,
+        name: XattrName,
+        value_reader: &mut VmReader,
+        flags: XattrSetFlags,
+    ) -> Result<()>;
+    fn get_xattr(&self, name: XattrName, value_writer: &mut VmWriter) -> Result<usize>;
+    fn list_xattr(&self, namespace: XattrNamespace, list_writer: &mut VmWriter) -> Result<usize>;
+    fn remove_xattr(&self, name: XattrName) -> Result<()>;
+}
+
+/// The index of the layer of an `OverlayFS`.
+type LayerIdx = u8; // Currently only support 256 layers.
+
+/// A visitor used by `OverlayFS` that merges the objects
+/// from the upper layer and the lower layer.
+struct OverlayDirVisitor {
+    dir_map: BTreeMap<usize, (String, u64, InodeType)>,
+    dir_set: HashSet<String>,
+    whiteout_set: HashSet<String>,
+    visited_files: usize,
+    cur_layer: LayerIdx,
+}
+
+impl DirentVisitor for OverlayDirVisitor {
+    fn visit(&mut self, name: &str, fs_ino: u64, type_: InodeType, fs_offset: usize) -> Result<()> {
+        if fs_offset & 0xFF00_0000_0000_0000 != 0 {
+            return_errno_with_message!(Errno::EOVERFLOW, "fs offset overflow")
+        }
+        // Use the layer index embedded in higher bits of the fs offset as the overlay offset.
+        let offset = fs_offset | ((self.cur_layer as usize) << 56);
+
+        if fs_ino & 0xFF00_0000_0000_0000 != 0 {
+            return_errno_with_message!(Errno::EOVERFLOW, "fs ino overflow")
+        }
+        // Use the layer index embedded in higher bits of the fs ino as the overlay xino.
+        // XXX: Linux uses embedded fsid for unique ino. Should we follow the technique?
+        let xino = fs_ino | ((self.cur_layer as u64) << 56);
+
+        if self.dir_set.contains(name) || self.whiteout_set.contains(name) {
+            return Ok(());
+        }
+        if name.starts_with(WHITEOUT_PREFIX) {
+            self.whiteout_set
+                .insert(name[WHITEOUT_PREFIX_SIZE..].to_string());
+            return Ok(());
+        }
+
+        let name = name.to_string();
+        self.dir_set.insert(name.clone());
+
+        if name != "." && name != ".." {
+            self.visited_files += 1;
+        }
+
+        debug_assert!(!self.dir_map.contains_key(&offset));
+        let _ = self.dir_map.insert(offset, (name, xino, type_));
```

Comment:
The two special entries, "." and "..", might be outputted multiple times because the stacked directories at different layers has their own "." and ".." entries and they won't be filtered by the `offset` argument of `readdir`. In addition, their inode numbers are incorrect, I believe.

Reply:
The filtering of "." and ".." entries is performed earlier in the `dir_set.contains(name)` check.